File: src/core/backup.ts
/**
 * Backup Management
 * Handles backup listing, restoration, cleanup, and rollback state
 */

import { readdir, stat, unlink, rename } from "node:fs/promises";
import { resolve, dirname, basename } from "node:path";
import type { BackupEntry, LastRunState, BackupConfig, InstallOptions } from "../types";
import { logger } from "./logger";
import { getHomeDir, contractPath } from "./os";

/** Pattern for backup files: filename.backup.timestamp */
const BACKUP_PATTERN = /^(.+)\.backup\.(\d+)$/;

/** Path to the last run state file */
const LAST_RUN_FILE = ".dotfiles-last-run.json";

/**
 * Get the path to the last run state file
 */
function getLastRunPath(): string {
  return resolve(getHomeDir(), LAST_RUN_FILE);
}

/**
 * Save the last run state
 */
export async function saveLastRunState(state: LastRunState): Promise<void> {
  const path = getLastRunPath();
  await Bun.write(path, JSON.stringify(state, null, 2));
  logger.debug(`Saved last run state to ${contractPath(path)}`, true);
}

/**
 * Load the last run state
 */
export async function loadLastRunState(): Promise<LastRunState | null> {
  const path = getLastRunPath();
  const file = Bun.file(path);

  if (!(await file.exists())) {
    return null;
  }

  try {
    const content = await file.text();
    return JSON.parse(content) as LastRunState;
  } catch (error) {
    logger.warn(`Failed to parse last run state: ${error}`);
    return null;
  }
}

/**
 * Find all backup files in a directory
 */
async function findBackupsInDir(dir: string): Promise<BackupEntry[]> {
  const backups: BackupEntry[] = [];

  try {
    const entries = await readdir(dir);

    for (const entry of entries) {
      const match = entry.match(BACKUP_PATTERN);
      if (match) {
        const [, originalName, timestampStr] = match;
        const timestamp = parseInt(timestampStr, 10);
        const backupPath = resolve(dir, entry);
        const originalPath = resolve(dir, originalName);

        backups.push({
          original: originalPath,
          backup: backupPath,
          timestamp,
        });
      }
    }
  } catch {
    // Directory doesn't exist or can't be read
  }

  return backups;
}

/**
 * Find all backup files across common dotfile locations
 */
export async function findAllBackups(): Promise<BackupEntry[]> {
  const homeDir = getHomeDir();
  const allBackups: BackupEntry[] = [];

  // Directories to search for backups
  const dirsToSearch = [
    homeDir,
    resolve(homeDir, ".config"),
    resolve(homeDir, ".config/starship"),
    resolve(homeDir, ".config/ghostty"),
    resolve(homeDir, ".config/kitty"),
    resolve(homeDir, ".claude"),
  ];

  for (const dir of dirsToSearch) {
    const backups = await findBackupsInDir(dir);
    allBackups.push(...backups);
  }

  // Sort by timestamp (newest first)
  allBackups.sort((a, b) => b.timestamp - a.timestamp);

  return allBackups;
}

/**
 * List all backup files
 */
export async function listBackups(): Promise<void> {
  const backups = await findAllBackups();

  if (backups.length === 0) {
    logger.info("No backup files found.");
    return;
  }

  logger.header("Backup Files");

  for (const backup of backups) {
    const date = new Date(backup.timestamp).toLocaleString();
    console.log(`  ${contractPath(backup.backup)}`);
    console.log(`    Original: ${contractPath(backup.original)}`);
    console.log(`    Created:  ${date}`);
    console.log();
  }

  logger.info(`Total: ${backups.length} backup(s)`);
}

/**
 * Restore a specific backup
 */
export async function restoreBackup(
  backupPath: string,
  options: InstallOptions
): Promise<boolean> {
  const expandedPath = backupPath.startsWith("~")
    ? resolve(getHomeDir(), backupPath.slice(2))
    : resolve(backupPath);

  // Parse backup filename to get original name
  const filename = basename(expandedPath);
  const match = filename.match(BACKUP_PATTERN);

  if (!match) {
    logger.error(`Not a valid backup file: ${backupPath}`);
    return false;
  }

  const [, originalName] = match;
  const originalPath = resolve(dirname(expandedPath), originalName);

  // Check if backup exists
  try {
    await stat(expandedPath);
  } catch {
    logger.error(`Backup file not found: ${contractPath(expandedPath)}`);
    return false;
  }

  if (options.dryRun) {
    logger.dryRun(`Would restore ${contractPath(expandedPath)} to ${contractPath(originalPath)}`);
    return true;
  }

  // Remove current file if it exists
  try {
    await unlink(originalPath);
  } catch {
    // File doesn't exist, that's fine
  }

  // Restore backup
  await rename(expandedPath, originalPath);
  logger.success(`Restored ${contractPath(originalPath)} from backup`);

  return true;
}

/**
 * Clean old backups based on retention policy
 */
export async function cleanBackups(
  config: BackupConfig,
  options: InstallOptions
): Promise<number> {
  const backups = await findAllBackups();
  const now = Date.now();
  const maxAgeMs = config.maxAge * 24 * 60 * 60 * 1000;

  // Group backups by original file
  const byOriginal = new Map<string, BackupEntry[]>();
  for (const backup of backups) {
    const existing = byOriginal.get(backup.original) ?? [];
    existing.push(backup);
    byOriginal.set(backup.original, existing);
  }

  let removed = 0;

  for (const [original, fileBackups] of byOriginal) {
    // Sort by timestamp (newest first)
    fileBackups.sort((a, b) => b.timestamp - a.timestamp);

    for (let i = 0; i < fileBackups.length; i++) {
      const backup = fileBackups[i];
      const age = now - backup.timestamp;
      const isOld = age > maxAgeMs;
      const exceedsCount = i >= config.maxCount;

      if (isOld || exceedsCount) {
        if (options.dryRun) {
          logger.dryRun(`Would remove: ${contractPath(backup.backup)}`);
        } else {
          try {
            await unlink(backup.backup);
            logger.success(`Removed: ${contractPath(backup.backup)}`);
            removed++;
          } catch (error) {
            logger.error(`Failed to remove ${contractPath(backup.backup)}: ${error}`);
          }
        }
      }
    }
  }

  return removed;
}

/**
 * Rollback to the state before the last install/link
 */
export async function rollback(options: InstallOptions): Promise<boolean> {
  const lastRun = await loadLastRunState();

  if (!lastRun) {
    logger.error("No previous run state found. Cannot rollback.");
    return false;
  }

  logger.header("Rolling Back");
  logger.info(`Last run: ${lastRun.command} at ${lastRun.timestamp}`);

  // Remove symlinks that were created
  logger.subheader("Removing symlinks");
  for (const entry of lastRun.symlinks) {
    if (options.dryRun) {
      logger.dryRun(`Would remove symlink: ${contractPath(entry.target)}`);
    } else {
      try {
        await unlink(entry.target);
        logger.success(`Removed: ${contractPath(entry.target)}`);
      } catch {
        logger.skip(`${contractPath(entry.target)} (not found)`);
      }
    }
  }

  // Restore backups
  if (lastRun.backups.length > 0) {
    logger.subheader("Restoring backups");
    for (const backup of lastRun.backups) {
      if (options.dryRun) {
        logger.dryRun(`Would restore: ${contractPath(backup.original)}`);
      } else {
        try {
          await rename(backup.backup, backup.original);
          logger.success(`Restored: ${contractPath(backup.original)}`);
        } catch (error) {
          logger.error(`Failed to restore ${contractPath(backup.original)}: ${error}`);
        }
      }
    }
  }

  // Remove the last run state file
  if (!options.dryRun) {
    try {
      await unlink(getLastRunPath());
    } catch {
      // File doesn't exist, that's fine
    }
  }

  logger.newline();
  logger.success("Rollback complete!");
  return true;
}

File: internal/backup/backup.go
package backup

import (
	"encoding/json"
	"fmt"
	"path/filepath"
	"regexp"
	"sort"
	"time"

	"github.com/alexcatdad/paw/internal/config"
	"github.com/alexcatdad/paw/internal/hooks"
	"github.com/alexcatdad/paw/internal/output"
	"github.com/alexcatdad/paw/internal/repo"
	"github.com/alexcatdad/paw/internal/state"
)

var backupPattern = regexp.MustCompile(`^(.+)\.backup\.(\d+)$`)

type BackupEntry struct {
	Original  string `json:"original"`
	Backup    string `json:"backup"`
	Timestamp int64  `json:"timestamp"`
}

type SymlinkEntry struct {
	Source string `json:"source"`
	Target string `json:"target"`
}

type LastRunState struct {
	Timestamp string         `json:"timestamp"`
	Command   string         `json:"command"`
	Backups   []BackupEntry  `json:"backups"`
	Symlinks  []SymlinkEntry `json:"symlinks"`
}

type Options struct {
	DryRun    bool
	SkipHooks bool
}

func SaveLastRun(st LastRunState) error {
	path, err := state.LastRunPath()
	if err != nil {
		return err
	}
	payload, err := json.MarshalIndent(st, "", "  ")
	if err != nil {
		return err
	}
	return fsys.WriteFile(path, payload, 0o644)
}

func LoadLastRun() (*LastRunState, error) {
	paths := []func() (string, error){state.LastRunPath, state.LegacyLastRunPath}
	for _, fn := range paths {
		path, err := fn()
		if err != nil {
			continue
		}
		data, err := fsys.ReadFile(path)
		if err != nil {
			continue
		}
		var st LastRunState
		if err := json.Unmarshal(data, &st); err != nil {
			continue
		}
		return &st, nil
	}
	return nil, nil
}

func FindAll(homeDir string) ([]BackupEntry, error) {
	dirs := []string{
		homeDir,
		filepath.Join(homeDir, ".config"),
		filepath.Join(homeDir, ".claude"),
	}
	result := []BackupEntry{}
	for _, dir := range dirs {
		entries, err := fsys.ReadDir(dir)
		if err != nil {
			continue
		}
		for _, e := range entries {
			if e.IsDir() {
				continue
			}
			name := e.Name()
			match := backupPattern.FindStringSubmatch(name)
			if len(match) != 3 {
				continue
			}
			var stamp int64
			fmt.Sscanf(match[2], "%d", &stamp)
			result = append(result, BackupEntry{
				Original:  filepath.Join(dir, match[1]),
				Backup:    filepath.Join(dir, name),
				Timestamp: stamp,
			})
		}
	}
	sort.Slice(result, func(i, j int) bool {
		return result[i].Timestamp > result[j].Timestamp
	})
	return result, nil
}

func List(logger *output.Logger) error {
	home, err := repo.HomeDir()
	if err != nil {
		return err
	}
	backups, err := FindAll(home)
	if err != nil {
		return err
	}
	if len(backups) == 0 {
		logger.Info("No backup files found")
		return nil
	}
	logger.Header("Backup Files")
	for _, b := range backups {
		logger.Info(repo.ContractHome(b.Backup))
	}
	return nil
}

func Restore(path string, dryRun bool, logger *output.Logger) error {
	home, err := repo.HomeDir()
	if err != nil {
		return err
	}
	resolved, err := repo.ExpandHome(path)
	if err != nil {
		return err
	}
	if !filepath.IsAbs(resolved) {
		resolved = filepath.Clean(resolved)
	}
	if err := repo.ValidateWithinBase(resolved, home); err != nil {
		return err
	}
	name := filepath.Base(resolved)
	match := backupPattern.FindStringSubmatch(name)
	if len(match) != 3 {
		return fmt.Errorf("invalid backup file name: %s", name)
	}
	original := filepath.Join(filepath.Dir(resolved), match[1])
	if dryRun {
		logger.DryRun(fmt.Sprintf("Would restore %s", repo.ContractHome(original)))
		return nil
	}
	_ = fsys.Remove(original)
	if err := fsys.Rename(resolved, original); err != nil {
		return err
	}
	logger.Success(fmt.Sprintf("Restored %s", repo.ContractHome(original)))
	return nil
}

func Clean(cfg config.BackupConfig, dryRun bool, logger *output.Logger) (int, error) {
	home, err := repo.HomeDir()
	if err != nil {
		return 0, err
	}
	backups, err := FindAll(home)
	if err != nil {
		return 0, err
	}
	cutoff := clk.Now().Add(-time.Duration(cfg.MaxAge) * 24 * time.Hour).UnixMilli()
	grouped := map[string][]BackupEntry{}
	for _, b := range backups {
		grouped[b.Original] = append(grouped[b.Original], b)
	}
	removed := 0
	for _, perFile := range grouped {
		sort.Slice(perFile, func(i, j int) bool { return perFile[i].Timestamp > perFile[j].Timestamp })
		for idx, item := range perFile {
			if idx < cfg.MaxCount && item.Timestamp >= cutoff {
				continue
			}
			if dryRun {
				logger.DryRun(fmt.Sprintf("Would remove %s", repo.ContractHome(item.Backup)))
				continue
			}
			if err := fsys.Remove(item.Backup); err == nil {
				removed++
			}
		}
	}
	return removed, nil
}

func Rollback(cfg config.Config, opts Options, logger *output.Logger) error {
	if err := hooks.Run("pre_rollback", cfg, hooks.Options{DryRun: opts.DryRun, SkipHooks: opts.SkipHooks}, logger); err != nil {
		return err
	}
	lastRun, err := LoadLastRun()
	if err != nil {
		return err
	}
	if lastRun == nil {
		return fmt.Errorf("no previous run state found")
	}
	for _, link := range lastRun.Symlinks {
		if opts.DryRun {
			logger.DryRun(fmt.Sprintf("Would remove symlink %s", repo.ContractHome(link.Target)))
			continue
		}
		_ = fsys.Remove(link.Target)
	}
	for _, backup := range lastRun.Backups {
		if opts.DryRun {
			logger.DryRun(fmt.Sprintf("Would restore %s", repo.ContractHome(backup.Original)))
			continue
		}
		_ = fsys.Remove(backup.Original)
		if err := fsys.Rename(backup.Backup, backup.Original); err != nil {
			logger.Warn(fmt.Sprintf("Failed to restore %s: %v", repo.ContractHome(backup.Original), err))
		}
	}
	if !opts.DryRun {
		if path, err := state.LastRunPath(); err == nil {
			_ = fsys.Remove(path)
		}
	}
	if err := hooks.Run("post_rollback", cfg, hooks.Options{DryRun: opts.DryRun, SkipHooks: opts.SkipHooks}, logger); err != nil {
		return err
	}
	logger.Success("Rollback complete")
	return nil
}
